### PR TITLE
EVG-15423: remove double slashes in URL to prevent open redirect issues

### PR DESCRIPTION
**Pull request by hhoke**
Double slash seems to have been causing an open redirect issue, see ticket for details. We now make sure that the path does not have any leading //, which would cause certain URLs with special characters to redirect to an external site.

### CODE REVIEW

**Comment by bsamek** [comment]
I can't really review this without an understanding of the problem and how this fixes it. Can @khelif96 review, or could you add some explanation?

**Comment by hhoke** [comment]
Per zoom comment: why not use `strings.ReplaceAll(r.URL.Path, "//", "/")`?

In the case where there are multiple slashes, you need a lookback method like the one I implement.
```
package main

import (
	"fmt"
	"strings"
)

func dedupeURLSlashes(path string) string {
        goodURL := ""
        lastChar := rune(0)
        for _, char := range path {
                if lastChar == char && lastChar == '/' {
                        continue
                }
                goodURL += string(char)
                lastChar = char
        }
        return goodURL
}

func main() {
	testStr := "this//that///and////another/////thing"
	fmt.Println(strings.ReplaceAll(testStr,"//","/"))
	fmt.Println(dedupeURLSlashes(testStr))
}
```
output:
```
this/that//and//another///thing
this/that/and/another/thing
```

File: vendor/github.com/urfave/negroni/static.go
Comment (khelif96) [inline vendor/github.com/urfave/negroni/static.go:104]:
```suggestion
return strings.ReplaceAll(path, "//", "/")
```

File: vendor/github.com/urfave/negroni/static.go
Comment (khelif96) [inline vendor/github.com/urfave/negroni/static.go:69]: The url spec seems to indicate that Path should never have a  preceding`//`
https://www.ietf.org/rfc/rfc3986.txt

> If a URI contains an authority component, then the path component
must either be empty or begin with a slash ("/") character. If a URI does not contain an authority component, then the path cannot begin
with two slash characters ("//"). In addition, a URI reference
(Section 4.1) may be a relative-path reference, in which case the
first path segment cannot contain a colon (":") character. The ABNF
requires five separate rules to disambiguate these cases, only one of which will match the path substring within a given URI reference. We use the generic term "path component" to describe the URI substring
matched by the parser to one of these rules.

      path          = path-abempty    ; begins with "/" or is empty
                    / path-absolute   ; begins with "/" but not "//"
                    / path-noscheme   ; begins with a non-colon segment
                    / path-rootless   ; begins with a segment
                    / path-empty      ; zero characters

I wonder if this should apply to the entire path. `dedupeURLSlashes` removes double `/` in the entire path. Which makes sense to me but it could be inadvertently breaking url behavior and potentially goes against the URL spec. 

for example 
/version/x/y/z
and 
/version/x//z 
are technically both valid urls.

File: vendor/github.com/urfave/negroni/static.go
Comment (khelif96) [reply vendor/github.com/urfave/negroni/static.go:72]: and only return the relative Path instead of the entire URL

**Review by khelif96 — CHANGES_REQUESTED** [review]
The concept makes sense to me but i had some suggestions.

File: vendor/github.com/urfave/negroni/static.go
Comment (hhoke) [reply vendor/github.com/urfave/negroni/static.go:69]: https://www.ctrl.blog/entry/nginx-requesturi-proxycache.html
This person seems to think it's non-standards-compliant. I'll only merge the first repeat.

**Comment by hhoke** [comment]
I implemented these suggestions in a PR to the main negroni repo. Hopefully the PR is accepted and this will save us from having to fork and revendor, or modifying here.
https://github.com/urfave/negroni/pull/259

**Review by bsamek — APPROVED** [review]
LGTM, but could you validate in staging if you haven't yet, and please wait for @khelif96 's re-review.

File: vendor/github.com/urfave/negroni/static.go
Comment (khelif96) [inline vendor/github.com/urfave/negroni/static.go:69]: Could we add a comment about why we do this for future reference.

**Review by khelif96 — APPROVED** [review]
Changes LGTM and make sense.

Seconded on validating in staging before merging. 

Also the test failures seem to have been introduced in this pr. I'm unsure of why the tests are failing but they seem to be responsible for how permissions are handled in evergreen. So we would want to ensure they are fixed before merging.

**Comment by hhoke** [comment]
I reran the tests and they now mysteriously work.

**Comment by hhoke** [comment]
I predict the next set of tests will mysteriously fail, then work once I rerun them

**Comment by hhoke** [comment]
Tested on staging. No errors in splunk, open redirect does not occur, patch introduces no new test failures https://spruce.mongodb.com/version/61530969d6d80a2543012275/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

**Pull request merged**
